Feat/prsd 364 better error handling

AEPR opened the pull request:
Added a couple of custom exceptions to allow the NotifyService to throw depending on the context. I've made a base exception class intended for the whole of the project so if we want to do a top level catch on that at some point we can.

I'm not entirely sure about the amount of content I've ended up putting into the NotifyService here either:
- All those classes to deserialise the error messages into a useful format plus the helper method to strip out the non-json bit.
- The actual exception message content for the 3 cases I've separated out.
On the other hand, they are private to the service so it makes sense that they are inside it! Let me know what you think.

I've also updated the "Send test email" page to take advantage of this new functionality so that if a temporary or one off error it reloads the page with an updated title. This is going to be hard to test/simulate as you would need to actually cause an internal server error or send enough emails to hit rate limiting.

rowanhill commented on src/main/kotlin/uk/gov/communities/prsdb/webapp/services/NotifyEmailNotificationService.kt:
```diff
@@ -13,6 +19,62 @@ class NotifyEmailNotificationService<EmailModel : EmailTemplateModel>(
+        var jsonString = nonJsonRegex.replace(message, "")
+        var parsed = Json.decodeFromString<NotifyErrors>(jsonString)
+
+        return parsed.errors.first().error
```
Without being explicit, the documentation _suggests_ we'll only get one error code back: https://docs.notifications.service.gov.uk/java.html#send-an-email-error-codes

Perhaps we could ask in #govuk-notify in the UK Government Digital slack.

AEPR replied:
I don't have access there, so yes please. I have indeed taken it as implicit that there will only ever be one error here - but obviously using first not single means it doesn't error otherwise so definitely worth checking.

AEPR replied:
I've updated the class to handle multiple errors more gracefully. It will throw the exception matching the "most severe" error (Auth>Bad Request>Daily Limit>Rate Limit>Exception), but if there were multiple errors there's a prefix to the error message saying that that's the case.

I don't think it's necessary to try and extract all the information as the original exception is available as a "Cause".

AEPR commented on src/main/kotlin/uk/gov/communities/prsdb/webapp/exceptions/CannotSendEmailsException.kt:
```diff
@@ -0,0 +1,6 @@
+package uk.gov.communities.prsdb.webapp.exceptions
+
+class CannotSendEmailsException(
```
Yes, I agree - I struggled a bit to come up with good names for these that express the idea of "That email wasn't sent but no comment on other emails" vs "We are currently unable to send emails and don't expect that to change imminently". I might open it up to the team for suggestions.

AEPR replied:
Taken Rowan's suggestions of "[Permanent/Transient] Email Send Exception"

AEPR commented:
> On where the serialization lives - I think that makes sense in the service as a it's specific to Notify. Related to that (but v. minor) we probably want our more generic email exceptions to not reference notify in their error messages as in theory we'd want to reuse them if we had to swap out to a different email service, for example.

I think it's fine for the error messages to reference notify in the string literals that come from the notify service. If we re-use the exceptions in a different email sending service they can have different error messages when they're thrown from those services.

AEPR commented on src/main/kotlin/uk/gov/communities/prsdb/webapp/services/NotifyEmailNotificationService.kt:
```diff
@@ -13,6 +19,62 @@ class NotifyEmailNotificationService<EmailModel : EmailTemplateModel>(
+            val errorType = parseNotifyExceptionError(notifyException.message ?: "")
+            when (errorType) {
+                NotifyErrorType.AUTH, NotifyErrorType.BAD_REQUEST -> throw CannotSendEmailsException(
+                    "prsdb-web does not have the correct credentials to send emails with Notify." +
```
Status code: 400 {"errors":[{"error":"BadRequestError","message":"Template not found"}],"status_code":400}

So I should change the error message to be less specific in any case.

AEPR replied:
Yeah, I've actually decided to split out all of the different cases - if we've hit our rate limit it's probably useful to have a different message than if Notify has had an internal server error.

AEPR commented:
> > I think it's fine for the error messages to reference notify in the string literals that come from the notify service. If we re-use the exceptions in a different email sending service they can have different error messages when they're thrown from those services.
> 
> Yes agree - but in one case we refer to it in the reusable exception directly - I've added a comment there as well to point it out in case it was a mistake - but let me know if I'm misunderstanding what you mean here!

Yes, I realised that I'd done this when I went back to start amending the other error messages. I've changed it so all messages are set in the service now.

AEPR commented on src/main/kotlin/uk/gov/communities/prsdb/webapp/exceptions/EmailWasNotSentException.kt:
```diff
@@ -0,0 +1,5 @@
+
+class EmailWasNotSentException(
+    cause: Throwable,
+) : PrsdbWebException("Notify has not sent that email, but retrying may work.", cause)
```
Moved it to the service.
